pkg/kubefed2/disable.go

```diff
@@ -139,75 +167,184 @@ func (j *disableType) Run(cmdOut io.Writer, config util.FedConfig) error {
-	err = client.Get(context.TODO(), typeConfig, typeConfigName.Namespace, typeConfigName.Name)
+	ftcExists, err := checkFederatedTypeConfigExists(client, typeConfig, typeConfigName, write)
 	if err != nil {
-		return errors.Wrapf(err, "Error retrieving FederatedTypeConfig %q", typeConfigName)
+		return err
 	}
 
 	if dryRun {
 		return nil
 	}
 
-	write := func(data string) {
-		if cmdOut == nil {
-			return
+	if deleteCRD {
+		if !ftcExists {
+			typeConfig, err = generatedFederatedTypeConfig(config, enableTypeDirective)
+			if err != nil {
+				return err
+			}
 		}
+		err = deleteFederatedType(config, typeConfig, write)
+		if err != nil {
+			return err
+		}
+	}
 
-		if _, err := cmdOut.Write([]byte(data)); err != nil {
-			glog.Fatalf("Unexpected err: %v\n", err)
+	if ftcExists {
+		if typeConfig.Spec.PropagationEnabled {
+			err = disablePropagation(client, typeConfig, typeConfigName, write)
+			if err != nil {
+				return err
+			}
+		}
+		checkPropagationControllerStopped(client, typeConfigName, write)
+		err = deleteFederatedTypeConfig(client, typeConfig, typeConfigName, write)
+		if err != nil {
+			return err
 		}
 	}
 
+	return nil
+}
+
+func checkFederatedTypeConfigExists(client genericclient.Client, typeConfig *fedv1a1.FederatedTypeConfig, typeConfigName ctlutil.QualifiedName, write func(string)) (bool, error) {
+	err := client.Get(context.TODO(), typeConfig, typeConfigName.Namespace, typeConfigName.Name)
+	if err == nil {
+		return true, nil
+	}
+
+	if apierrors.IsNotFound(err) {
+		write(fmt.Sprintf("FederatedTypeConfig %q does not exist\n", typeConfigName))
+		return false, nil
+	}
+
+	return false, errors.Wrapf(err, "Error retrieving FederatedTypeConfig %q", typeConfigName)
+}
+
+func disablePropagation(client genericclient.Client, typeConfig *fedv1a1.FederatedTypeConfig, typeConfigName ctlutil.QualifiedName, write func(string)) error {
 	if typeConfig.Spec.PropagationEnabled {
 		typeConfig.Spec.PropagationEnabled = false
-		err = client.Update(context.TODO(), typeConfig)
+		err := client.Update(context.TODO(), typeConfig)
 		if err != nil {
 			return errors.Wrapf(err, "Error disabling propagation for FederatedTypeConfig %q", typeConfigName)
 		}
 		write(fmt.Sprintf("Disabled propagation for FederatedTypeConfig %q\n", typeConfigName))
 	} else {
 		write(fmt.Sprintf("Propagation already disabled for FederatedTypeConfig %q\n", typeConfigName))
 	}
-	if !delete {
-		return nil
-	}
+	return nil
+}
+
+func checkPropagationControllerStopped(client genericclient.Client, typeConfigName ctlutil.QualifiedName, write func(string)) {
+	write(fmt.Sprintf("Checking propagation controller is stopped for FederatedTypeConfig %q\n", typeConfigName))
+
+	var typeConfig *fedv1a1.FederatedTypeConfig
+	err := wait.PollImmediate(100*time.Millisecond, 10*time.Second, func() (bool, error) {
+		typeConfig = &fedv1a1.FederatedTypeConfig{}
+		err := client.Get(context.TODO(), typeConfig, typeConfigName.Namespace, typeConfigName.Name)
+		if err != nil {
```

Comment:
Is there a good reason not to log this error?

Reply:
Avoiding dumping 100 errors to the console buffer in the event the error persists for the entire duration of the poll. I went ahead and added it since it's not in the normal path we expect.

Reply:
(No action required) Hmm, maybe we need a wait loop that does exponential backoff on error?